Updated to match json-api spec on one to many relationships

dylanswartz opened the pull request:
Hi friend! 

First of all thanks for building this, love it! Someone on my team ran into an issue when integrating our web service with our Ember.js App (2.x consumes JSON-API out of the box). We have a one to many relationship in our data model on both the server and client side. Turns out the waterline-jsonapi module was actually transforming this into a one to one representation when the collection only contained one element (on the many side of the association) causing Ember to freak out. 

We wrote a custom serializer on the ember-side for now as a workaround.  

I noticed your issue regarding this #1 however I think it may be a mis interpretation of the spec (I did that too at first). 

According to the JSON-API spec resource linkage MUST be represented as one of the following:
* null for empty to-one relationships.
* an empty array ([]) for empty to-many relationships.
* a single resource identifier object for non-empty to-one relationships.
* an array of resource identifier objects for non-empty to-many relationships.

Notice bullet 3 "a single resource identifier object for non-empty to-one relationships", that's for non empty *to-one* relationships. If we have an array I think it's safe to assume this is a *to-many* relationship. See: http://jsonapi.org/format/#document-resource-object-related-resource-links

Thanks! 
Dylan

davemackintosh commented:
Good spot! Thanks for this, looks good to me 'part from my one comment on styling. Will merge once you've made that small change, could you squash your commits too please?

I'll get this published today :hand:

dylanswartz commented:
@davemackintosh I've refactored the map_target_values function as requested and squashed the commits! 

I also noticed an edge case in where if `values` was undefined `values.is_error` was causing my app to crash. I'm not quite sure exactly how you'd like to handle this so I just added a conditional at the top: 
https://github.com/MasonAmerica/waterline-jsonapi/blob/master/lib/data.js#L18

davemackintosh commented:
Undefined and/or null? What's the expected output and what does the conditional make the output?

dylanswartz commented:
@davemackintosh Apologies for the delay. Don't seem to be getting notifications from Github. Here's the error I'm getting every so often in my app: 

2016-01-23T18:28:34.174317+00:00 app[web.1]: /app/node_modules/waterline-jsonapi/lib/data.js:23
2016-01-23T18:28:34.174322+00:00 app[web.1]:   if (values instanceof Error || values.is_error) {
2016-01-23T18:28:34.174324+00:00 app[web.1]:                                        ^
2016-01-23T18:28:34.174326+00:00 app[web.1]:
2016-01-23T18:28:34.174334+00:00 app[web.1]:     at Object.new_from_values (/app/node_modules/waterline-jsonapi/index.js:54:14)
2016-01-23T18:28:34.174332+00:00 app[web.1]: TypeError: Cannot read property 'is_error' of undefined
2016-01-23T18:28:34.174334+00:00 app[web.1]:     at /app/server/api/clients.js:18:42
2016-01-23T18:28:34.174333+00:00 app[web.1]:     at [object Object].JSONAPIModel (/app/node_modules/waterline-jsonapi/lib/data.js:23:40)
2016-01-23T18:28:34.174337+00:00 app[web.1]:     at wrappedCallback (/app/node_modules/waterline/lib/waterline/utils/normalize.js:320:18)
2016-01-23T18:28:34.174335+00:00 app[web.1]:     at wrapper (/app/node_modules/waterline/node_modules/lodash/index.js:3602:19)
2016-01-23T18:28:34.174341+00:00 app[web.1]:     at wrapper (/app/node_modules/waterline/node_modules/lodash/index.js:3602:19)
2016-01-23T18:28:34.174336+00:00 app[web.1]:     at applyInOriginalCtx (/app/node_modules/waterline/lib/waterline/utils/normalize.js:421:80)
2016-01-23T18:28:34.174338+00:00 app[web.1]:     at callback.success (/app/node_modules/switchback/lib/normalize.js:33:31)
2016-01-23T18:28:34.174339+00:00 app[web.1]:     at returnResults (/app/node_modules/waterline/lib/waterline/query/finders/basic.js:180:9)
2016-01-23T18:28:34.174340+00:00 app[web.1]:     at /app/node_modules/waterline/lib/waterline/query/finders/basic.js:91:16
2016-01-23T18:28:34.174343+00:00 app[web.1]:     at wrappedCallback (/app/node_modules/waterline/lib/waterline/utils/normalize.js:320:18)
2016-01-23T18:28:34.174338+00:00 app[web.1]:     at _switch (/app/node_modules/switchback/lib/factory.js:58:28)
2016-01-23T18:28:34.198920+00:00 app[web.1]:
2016-01-23T18:28:34.174340+00:00 app[web.1]:     at /app/node_modules/waterline/lib/waterline/query/finders/operations.js:77:45
2016-01-23T18:28:34.174344+00:00 app[web.1]:     at callback.success (/app/node_modules/switchback/lib/normalize.js:33:31)
2016-01-23T18:28:34.174342+00:00 app[web.1]:     at applyInOriginalCtx (/app/node_modules/waterline/lib/waterline/utils/normalize.js:421:80)
2016-01-23T18:28:34.174344+00:00 app[web.1]:     at _switch (/app/node_modules/switchback/lib/factory.js:58:28)
2016-01-23T18:28:34.205335+00:00 app[web.1]: npm ERR! node v5.4.0
2016-01-23T18:28:34.214173+00:00 app[web.1]: npm ERR!     /app/npm-debug.log
2016-01-23T18:28:34.206284+00:00 app[web.1]: npm ERR! npm  v3.3.12
2016-01-23T18:28:34.205046+00:00 app[web.1]: npm ERR! argv "/app/.heroku/node/bin/node" "/app/.heroku/node/bin/npm" "start"
2016-01-23T18:28:34.213929+00:00 app[web.1]: npm ERR! Please include the following file with any support request:
2016-01-23T18:28:34.204514+00:00 app[web.1]: npm ERR! Linux 3.13.0-71-generic
2016-01-23T18:28:34.206918+00:00 app[web.1]: npm ERR! code ELIFECYCLE
2016-01-23T18:28:34.207193+00:00 app[web.1]: npm ERR! platform-service@0.0.9 start: `node index.js`
2016-01-23T18:28:34.207455+00:00 app[web.1]: npm ERR! Exit status 1

I'll take a deeper look into this and provide you with more information.

davemackintosh commented:
Ah okay, you're check is valid then. I'll merge this now and publish. Sorry for the delay, been crazy busy with Multicolour.
